### PR TITLE
Add Reading Groups API + Demo view

### PR DESCRIPTION
**Pull request by EfronC**
### Reading Groups API
Add a feature for Reading Groups to LRR. This is a feature to group archives into an ordered list, so it would be easier to read related works, like separate works from an Story arc.

The code is mainly the API backend code, the frontend files provided is just a simple demo to show an idea for how I made it to work, still I think there might be more efficient designs for the view. I tried to separate all these elements from the base code so that way would be easier to remove them if necessary. The Demo can be checked at: http://localhost:3000/rgroups . 

The demo doesn't include a way to add files to the Reading Groups, since the intended way for it to work was to add a button in the context menu of files in the search view, so you can add them directly from there. Said way was not implemented so I didn't have to touch the main code. For testing purposes I made a direct request to the API to add those files manually.

The Edit button in the menu of the Carousel opens a draggable list to reorder the elements, I implemented a simple one from a quick Google search, but I'm pretty sure there are more beautiful ones out there that probably would require new libraries.

PS: Just an aditional comment, The code is not "Complete", I left all endpoints public and didn't set pagination in the lists, for the demo to work.

### CODE REVIEW

**Review by Difegue — CHANGES_REQUESTED** [review]
Thanks a lot for submitting this, and sorry for taking so long to get to reviewing it!  

This feature is essentially what I'd drafted in #519, but I never wrote any code for it as it required the now-completed Search revamp and is a pretty heavy workload... I'd like to take this PR as an opportunity to actually start working towards it.   

I agree that UI for this is pretty tough to get right, but having an Edit page to create/delete Groups and reorder stuff in them feels like the bare minimum, so you were right to make that one imo!  
Playing around with your demo got me thinking about some UI concepts to add archives to groups/categories from the main index, I'll try to draft a spec for that soonish. 👍
 

I think what we could do here is try focusing towards **getting the API finished up + the Edit page if possible**? I've added a few comments to try and get us started. 🏃 
A fair bit of those are open questions since I'm not 100% on the design, so opinions welcome.

After that, what's left is search and reader integration:  

* hide IDs that are in a group and show the group in the index instead
* have the group come up when searching for tags that are in any of the included IDs
* having the reader be able to handle groups by opening the first ID, then the subsequent ones when done reading
* handling reading progress in a group
* being able to add groups to categories (not necessary, that'll probably come later imo)

I'll take over afterwards to write those, but if you want to keep helping I wouldn't refuse either 😅

File: lib/LANraragi/Controller/Api/Readinggroup.pm
Comment (EfronC) [reply lib/LANraragi/Controller/Api/Readinggroup.pm:1]: No, no, I just named it that way because it's the first thing that came to my mind in that moment(And ended up regretting it due to naming conventions of Mojo). If you tell me which you think would be the adequate name, then I'll rename all the modules.

Just in the topic, while Tanks would be the main use-case for this module, it's important to mention that right now it's possible to create lists for generic groups(Like "Pending to read") where the groups might not be of "Series" related works, but "Reader" related works, which also means a file could be included in more than one list, but if this sounds too complex, I can make it so an arcid can only be in one list at a time. Here I'll let you decide which sounds better, and what name would work the best for the module(In my opinion, Tankoubon sounds good).

File: lib/LANraragi/Utils/Routing.pm
Comment (EfronC) [reply lib/LANraragi/Utils/Routing.pm:155]: Yes, I left all public for the Demo to be easily testeable, but I'll change all the POST, UPDATE, and DELETE  EPs to login required when I commit the final version.

I can make an EP to return the groups where an Arcid is present, only that complexity could raise very quickly depending in how is used. For example, my main idea(As someone that hasn't worked a lot with Redis, so maybe I'm saying gibberish) was to create an exclusive table GROUPED which would contain a list of hashes, where a hash would contain a list of the group ids where is contained, so this way: Want to know if the work is contained somewhere? Check if the Arcid is in the list of hashes, Want to know which lists is contained? Access the key and that's all, that way the search would be, probably at most, O(n), otherwise to search in every list and between all the works would go to O(n*m). I'll check what can be done with Redis and update with my findings.

File: lib/LANraragi/Model/ReadingGroup.pm
Comment (EfronC) [reply lib/LANraragi/Model/ReadingGroup.pm:45]: If you point to a module that has that new style, I can refactor all my code to ensure it works similarly.

File: lib/LANraragi/Model/ReadingGroup.pm
Comment (EfronC) [reply lib/LANraragi/Model/ReadingGroup.pm:143]: Yep, because reordering is easier in the Frontend is that I made it this way, but an "Insert one" is maybe at least necessary, because that way you can insert easily a file from the search by getting only the id of the group instead of getting it's whole content(Although, I think that to get the name of the group I'm already returning it's content list).

File: lib/LANraragi/Controller/Api/Readinggroup.pm
Comment (EfronC) [reply lib/LANraragi/Controller/Api/Readinggroup.pm:98]: I thought the same at the start, but just to add to this point: In my way of think this, categories are probably more for massive amount of works(like all Non-H, or all CG style), kind of a custom super tag, while reading groups is more of like a way to group a limited number of works(Number that could get very big over the years depending on the use). I think the easier way to see it is that you're supposed to have a few categories, but every category contains a ton of works, while you can have a ton of groups, but every group contains just a few works. 

Might sound similiar, but those 2 are different methods of organization.

File: public/js/rgroups.js
Comment (EfronC) [reply public/js/rgroups.js:92]: Yeah, about the UI, I just want to mention that it's mostly a copy/paste for demo purposes, on my head every client developer would be making the UI whoever they thought would work the best, that's why the API tries to be generic instead of use-case specific. 

Still one aditional comment to add here, the way I thought this part would work(And I didn't add because I didn't want to keep fighting with the UI): First you open the list of groups(Which would be paginated and probably have a searcher), BUT every carrousel would have like 10 or 15 works, and at the end a "Check more" button, which then would take you to a new view, which would not be a carrousel, but probably a more Search like view where the list of works would be paginated, and that way would be easier to explore, here would also be reorder option, which I think will have problems anyway the bigger the number of works in the list, unless a "Move to n position" button is added, so that way if you added the 1000 chapters of OP, but forgot chapter 386, you can just add it at the end -> right click on it -> Move to position 386, and that way you don't have to move it manually either on the pop up nor the carrousel(But because all this is Frontend work, I don't want to opine a lot).

File: tools/build/docker/Makefile
Comment (EfronC) [reply tools/build/docker/Makefile:1]: Oh, yeah, this is a file that I forgot I commited, I'll remove it in the next commit.

File: lib/LANraragi/Utils/Routing.pm
Comment (EfronC) [reply lib/LANraragi/Utils/Routing.pm:89]: Which actually takes me a question I had when working on this: Should groups be user specific? I mean, suppose there are 3 persons using the platform, Should the 3 of them be looking at all the groups, or should everyone look only the lists they created? I think I'll add a public/private attribute anyway, just in case

File: lib/LANraragi/Model/ReadingGroup.pm
Comment (Difegue) [reply lib/LANraragi/Model/ReadingGroup.pm:143]: If we use sorted sets for the database structure it'd probably be easier to have insert/remove API endpoints at least -- No need to serialize/deserialize json every time 🥲

File: lib/LANraragi/Controller/Api/Readinggroup.pm
Comment (Difegue) [reply lib/LANraragi/Controller/Api/Readinggroup.pm:98]: Yeah now that I'm thinking about the different DB structures (+ the search implementation afterwards), it's probably better to keep categories and tanks separate as I believe the code will evolve in quite a different direction.

File: lib/LANraragi/Model/ReadingGroup.pm
Comment (Difegue) [reply lib/LANraragi/Model/ReadingGroup.pm:45]: Look at https://github.com/Difegue/LANraragi/blob/dev/lib/LANraragi/Utils/Archive.pm for an example.  
You'll have to declare 
```
use feature qw(signatures);
no warnings 'experimental::signatures';
``` 
in the opening of any file with sigs otherwise old Perls will yell at you

File: lib/LANraragi/Utils/Routing.pm
Comment (Difegue) [reply lib/LANraragi/Utils/Routing.pm:155]: Currently the equivalent endpoint for categories just iterates over every category in the DB and checks if the arcid is present in them. Its as inefficient as can be but I haven't received any complaints about it!  

Tanks could however have a simpler database structure than categories since they're just a name and a sorted list of IDs. (whereas categories have to bundle a search predicate, as well as pinned/last_used status)  

You might be able to make use of [Redis sorted sets](https://redis.io/docs/data-types/sorted-sets/) to save the ordered list of arcids (as `TANK_xxxxx` with the name in a separate `TANK_xxxxx_NAME` I guess?), at which point you could use stuff like `ZSCORE` to find the ID in each set, with each zscore being a O(1) operation.

File: lib/LANraragi/Controller/Api/Readinggroup.pm
Comment (Difegue) [reply lib/LANraragi/Controller/Api/Readinggroup.pm:1]: Let's go with Tankoubon then 👍 

My thought is that stuff like "pending to read" is more suited for categories, but considering how similar the two concepts are I wouldn't be surprised if people made that with tanks instead so they could reorder them.  

I don't think we should restrict arcid from being in more than one tank, that'd be a virtual limitation more than anything else 🤔

File: lib/LANraragi/Controller/Api/Readinggroup.pm
Comment (Difegue) [reply lib/LANraragi/Controller/Api/Readinggroup.pm:98]: (still think it's better to update the messages here to match the style of categories though, yeah! 😛)

File: lib/LANraragi/Utils/Routing.pm
Comment (Difegue) [reply lib/LANraragi/Utils/Routing.pm:89]: Well, technically there's only two possible users atm, logged in/admin or not logged in. I believe groups should be like categories and be accessible to all.. 

I've been thinking about adding a secondary password for reading/uploading for https://github.com/Difegue/LANraragi/issues/630 and https://github.com/Difegue/LANraragi/issues/740, but account-specific content is a bit out of that scope 🤔

File: public/js/rgroups.js
Comment (Difegue) [reply public/js/rgroups.js:92]: Yeah I was just throwing ideas around, fine to focus on the API only if you don't want to fight w/ javascript 👍  

Most reordering UI paradigms fall apart on large groups -- I'll probably have to add some emergency "move to position X" with a textbox as you mentioned.

File: lib/LANraragi/Utils/Routing.pm
Comment (EfronC) [reply lib/LANraragi/Utils/Routing.pm:155]: Ok, I've been testing the sorted sets for a while to check how to implement it, and seems doable, but there is a pair of problems(Which mostly are not something big, but are worth to mention):

First is that 'zrangebyscore' from Perl seems to not be doing what is supposed
image
Here in the first print and 3rd print you can see a call from this exact line `my %tankoubon = $redis->zrangebyscore($tank_id, 0, "+inf", "WITHSCORES");` but in one it returned the results ordered, and in the other reversed, this might be a problem, because it makes me require to add an aditional sort to make sure the list is ordered. Is not that big of a problem, but performance wise might eventually take a hit(Although, with this I can add pagination more easily with the LIMIT offset count option, so maybe is even better). BTW score 0 will be reserved for the name of the Tank, so no need to save it in the hash key.

Second is that, because I can't just overwrite what is saved in one, reorder and edit will probably require a little more work, since now I'll need to check which arcids are not more in what Front sends me to remove them with zrem, and manually update all the scores with zadd. I'm not sure if this could be a bad practice, but at the very least in the docs it says that those methods are O(log(N)) and O(M*log(N)), and I doubt someone will create a list with tens of thousands of files as to be notable, so I'm guessing it's ok. 

Will try to keep the EPs working the same, but maybe something that will change is that I'll add pagination in the list of Tanks and the list of archives, just to test.

**Comment by EfronC** [comment]
Some changes applied in most recent commit:

- Name of the module and all files was changed to Tankoubon.
- Tanks are not handled with Sorted Sets types in Redis, all modules in the model were modified to work with this. Body of the requests(From both way) was mostly left intact.
- Model subs now use signatures instead of '@_'.
- Success Message for Reorder was modified to match the one in categories.
- Get Tankoubon list, and Get Tankoubon Elements are now both paginated(Just add a page=<current> at the URI).

There might be some problems, since I just did simple tests with few files, so when the module is considered ready, will be recommendable to test with Tanks with a lot of elements.

**Comment by EfronC** [comment]
In this last commit, I add a simple method to search between all the Tanks, which contains the requested file. I added the method in Database.pm, so that way can be used in any place that requires this search. I also added a public Endpoint to use it, so that way the clients can do this themselves for extra features. 

I think with this I've already tackled all API related requested changes, so I'll keep in touch for any aditional commentary.

**Comment by EfronC** [comment]
Well, at last, a pair of tests for the get tank operations. Honestly I'm a bit dissapointed with the test module, since I thought would be a little more realistic, but well, at least I hope this 2 tests can eventually be helpful.

File: lib/LANraragi/Utils/Routing.pm
Comment (Difegue) [inline lib/LANraragi/Utils/Routing.pm:158]: In order to match the similar API for categories, maybe the path for this endpoint should be `/api/archives/:id/tankoubons` ?

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Model/Tankoubon.pm:258]: this sub can probably be trimmed down to just `return get_tankoubons_by_file($arcid);`

File: lib/LANraragi/Controller/Api/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Controller/Api/Tankoubon.pm:127]: not specified*

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Model/Tankoubon.pm:21]: Outdated comment since this sub now does have pagination?  

Speaking of pagination, since there's no real CPU gain here from adding it (we're still getting the entire list of tanks, just splitting it after the fact), maybe this sub should just return the full array, and the API sub should be the one to handle pagination?  

(I'd also advocate for maybe being able to pass page=-1 to grab everything w/o pagination which I think we're doing in at least one other API call somewhere in the app 🤔)

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Model/Tankoubon.pm:118]: Having one of the response fields be a completely different data type depending on a parameter is _hella weird_ API design imo -- I think it'd be better here to always include the ID list as `archives`, and have an additional `full_data` field present with the list of archive data if the API was called with the `decoded` parameter.  

I don't think `decoded` is a very good name either as it took me reading up to this implementation to figure out what it did; It's a bit verbose but maybe `include_full_data` would be better?

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Model/Tankoubon.pm:104]: If this endpoint has pagination, it should return some information as to how many IDs were returned and how many total pages/IDs exist.  

Similarly to the other endpoint, I think it should also have a way to disable pagination and just get all the IDs -- which wouldn't be too expensive unless you have a gigantic tankoubon and add `decoded` to the call.  
(at which point I'd say it's your fault as a client, what the fug are you doing)

File: lib/LANraragi/Controller/Api/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Controller/Api/Tankoubon.pm:30]: Are those default values required here considering the sub signature for the Model also has defaults?  

Ditto for get_list above.

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Model/Tankoubon.pm:64]: According to the comment on top this sub should just update the name if given an existing tank ID, but instead it's generating a new ID?

File: lib/LANraragi/Utils/Routing.pm
Comment (Difegue) [inline lib/LANraragi/Utils/Routing.pm:162]: Should we add `delete /api/tankoubon/:id/:archive` as well now?

**Review by Difegue — CHANGES_REQUESTED** [review]
Nice progress! Thanks for taking the time to write tests as well, even if they're simple for now.  

Here are some more suggestions from me.  
The API seems good bar a few things, so I think you could also update the [Documentation](https://github.com/Difegue/LANraragi/tree/dev/tools/Documentation/api-documentation) with the new endpoints? 🙏

File: lib/LANraragi/Utils/Routing.pm
Comment (Difegue) [reply lib/LANraragi/Utils/Routing.pm]: Looking back at this comment -- Do you still have the first problem? zrange and friends shouldn't reverse the order every time you call them!  

As for the second one, imo we could change the API to just have add/delete endpoints (with an optional position argument for add) and hand off all the complexity of knowing which archives to add/remove to the frontend/clients... 
The issue with that is that I don't think there's an easy way to update the scores of every ID in a sorted set when we move/insert a new one 🤔

Comment (EfronC) [reply lib/LANraragi/Utils/Routing.pm]: Hi, sorry for the lack of answers, I'm in the middle of a vacation, on Monday that I return I'll check the changes you requested, but for the moment I'll answer this one:

The 1st problem still happens, but as mentioned I solved it by manually sorting after the zrangebyscore, I'm not sure why is it giving me unordered data, but it happens at least on my machine, if someone can verify that is working correctly, then the loop marked with `# Sort and add IDs to archives array` in get_tankoubon can be removed(With the required changes to make the next lines keep working ofc).

For the 2nd, I already implemented those 2 methods in the update to make sure to apply all the insert/deletes in one in the update_archive_list with these lines

image

Basically, what I do is to check the left difference with the original, and arcids not anymore are removed with zrem, while with the new list I just regenerate all the scores(Again, for this I'm taking the array that is sent from Frontend is necessarily ordered) and update it that way. Simple and it works.

The problem with having add/delete endpoints is that if you want to add it at an specific place, you need to regenerate all the scores anyway, for example, if you have a 100 elements list, and you want to insert in position 4, from the actual 4 onward you would need to add 1 to the score and update, same case with the deletes, otherwise the list might end up messed up eventually with dupped scores. I still added an add_to_tankoubon method, but it adds it at the final position of the tank, since I can use zcard to get the size of the set to know its score. 

Also, just to reminder that another of the problems is the number of requests you would need to do to the back, might not be very bad, but sometimes might be better 1 good request than 100 smaller ones.

Comment (Difegue) [reply lib/LANraragi/Utils/Routing.pm]: No real rush! I'm going to be away next week as well so you can take your time. 👍  

I'll try removing the loop when running this locally at some point to verify - although if it's out of order on _some machines_ we'd better keep the sorting code in anyway..
I agree with the score regen being an unfixable problem (or at least idk how to fix it), so if we need to rescore everything might as well keep the current approach.

File: lib/LANraragi/Model/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Model/Tankoubon.pm:268]: Checked it, but from what I saw, your intersect returns either the intersect or full diff, mine returns only left diff, which is what I need because I'm comparing the original list of archives with the new, and the ones in the original that are not anymore in the new needs to be processed by zrem. Moved array_difference to Utils::Generic anyways.

File: lib/LANraragi/Utils/Routing.pm
Comment (EfronC) [reply lib/LANraragi/Utils/Routing.pm:158]: Renamed it, and moved in Routing to Archives API(Still the method is inside Tankoubon.pm).

File: lib/LANraragi/Model/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Model/Tankoubon.pm:21]: Well, is not that I added for the CPU gain(Which is very relevant BTW, I think Scan probably has an option to limit the results of the keys, but I'm not sure how to use it), but because is not a good practice to return lists of elements that can growth very large in a response, mostly because the response could one day get large enough to cause a timeout.

Still, I added the page=-1 so it can be tested how it works to have it return everything at once.

File: lib/LANraragi/Model/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Model/Tankoubon.pm:118]: Yes, I know it's weird, but disgracefully I've been in projects where we have to do it so we don't need to make redundants EPs... Anyway, I changed the name of the param to include_full_data, and like you said, it now appends a full_data field with the Archive objects.

File: lib/LANraragi/Model/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Model/Tankoubon.pm:64]: Not sure what you mean, but in this case I noted I didn't implement the edit with this(I was sending always tank_id as ""), so I added a few lines to edit the name.

File: lib/LANraragi/Utils/Routing.pm
Comment (EfronC) [reply lib/LANraragi/Utils/Routing.pm:162]: I still insist is not a good Idea to have individual remove, but I still added it anyway. It actually gets the score of the archive to remove, and modifiy everyone after it to reduce its score by 1. In case that is wanted that the add to to tank can insert on an specific position, it would need to do something similar, but changing the -1 to +1. 

Use it with wisdom.

File: lib/LANraragi/Controller/Api/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Controller/Api/Tankoubon.pm:30]: I removed them, but now the console throws a lot of warnings with `Use of uninitialized value $page in multiplication (*) at /home/koyomi/lanraragi/script/../lib/LANraragi/Model/Tankoubon.pm`, doesn't cause any error, but it's probably annoying for who is trying to debug. Not sure how to fix that.

**Comment by EfronC** [comment]
In the most recent commit I tackled many of the Change requests mentioned:

- I modified the Documentation to add Tankoubon API, but just to mention that since I wasn't able to preview what I was writting with Gitbook, I'm not sure if the syntax is correct. Also a few EPs might be outdated, since I was editing as I was fixing a few things from this commit, so the result may differ a little in maybe one or more cases.
- I changed Get Tankoubon to include Pagination data, and decoded now is called include_full_data, and appends a new field called full_data, which now will contain the Archive Objects.
- Pagination in Get Tankoubon and Get Tankoubon List now accepts page=-1 to return everything.
- Added the Endpoint Remove Archive from Tankoubon.
- Updated the routes for the Tankoubon API to have tankoubon in plural.
- Array_difference is now in Utils::Generic.pm
- Few aditional requested changes.

File: lib/LANraragi/Utils/Routing.pm
Comment (Difegue) [reply lib/LANraragi/Utils/Routing.pm]: Yeah, not sure I'll use it in the web UI, but I think it's worth having for clients with an extra warning label on top. Thanks!

File: lib/LANraragi/Utils/Routing.pm
Comment (Difegue) [reply lib/LANraragi/Utils/Routing.pm]: Eh, fine to leave the method in the Tank model I guess yeah ¯\\\_(ツ)_/¯

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Model/Tankoubon.pm:335]: Well, it'll just return an empty array in that case right

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [reply lib/LANraragi/Model/Tankoubon.pm]: Ah well, there's nothing disgraceful about being forced to implement bad upper management decisions imo 😤  
Thanks!

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [inline lib/LANraragi/Model/Tankoubon.pm:139]:
```suggestion
        eval { $tank{full_data} = \@data }
```

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [reply lib/LANraragi/Model/Tankoubon.pm:69]: Ah yeah my mistake, I didn't read the `unless` statement properly, it'd only generate a new tank if tank_id was blank.  
It was still missing the edit/name update as mentioned though, yep!

File: lib/LANraragi/Controller/Api/Tankoubon.pm
Comment (Difegue) [reply lib/LANraragi/Controller/Api/Tankoubon.pm]: Looked at the docs for sub signatures again, the default values are only applied if you don't give it a parameter, but in this case we're going to give them `undef`..  

But that can be fixed by using `//=` in the signature: 
> A default value expression can also be written using the //= operator, where it will be evaluated and used if the caller omitted a value or the value provided was undef.

```
sub foo ($name //= "world") {
    print "Hello, $name";
}

foo(undef);  # will print "Hello, world"
```

**Review by Difegue — CHANGES_REQUESTED** [review]
Thanks again for the documentation, it looks good to me at a glance. We'll see what gitbook makes of it once merged, sadly there's no way to test it in PRs. 🤔  

I've only got a few cleanup suggestions left + what seems to be a missing `zadd` in the edit function.

File: lib/LANraragi/Controller/Api/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Controller/Api/Tankoubon.pm]: Tried this, but trying to use //= in the signature throwed errors... I'm guessing that what you said should work, but maybe hasn't been implemented correctly(Or maybe I didn't), anyways, just to supress that warning for the moment I added a `$page //= 0 ` inside those subs, which is basically what I did at the start, but I guess it works until signatures gets out of experimental phase.

File: lib/LANraragi/Model/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Model/Tankoubon.pm]: Yeah, that's what I was doing, it returns () if the arcid does not exist. Guess there is no problem with that.

File: lib/LANraragi/Model/Tankoubon.pm
Comment (EfronC) [reply lib/LANraragi/Model/Tankoubon.pm:83]: There is, it's outside the else, because setting the name is an step that has to be done even if the list is new. The else is just to delete the old one(And I changed the name to @old_name).

**Comment by EfronC** [comment]
Ok, most recent one apply most of the cleanings requested :+1:

File: lib/LANraragi/Model/Tankoubon.pm
Comment (Difegue) [reply lib/LANraragi/Model/Tankoubon.pm:83]: oh woops it was hidden in the previous diff by github my bad

File: lib/LANraragi/Controller/Api/Tankoubon.pm
Comment (Difegue) [reply lib/LANraragi/Controller/Api/Tankoubon.pm]: Ah, maybe that depends on perl versions, yeah you did the right thing by adding the default assignation back!

**Review by Difegue — APPROVED** [review]

**Comment by Difegue** [comment]
Alright, let's get this first part merged!  
I think I've said it like 10 times already but thanks a lot for putting up with the whole thing, this really helps a lot. 💪  

I won't really have time to work on the rest until october due to other projects/obligations, so if you want to proceed with the search index/reading progress integrations, feel free. No obligations tho ofc

@holopin-bot @EfronC give-him-the-burg

**Pull request merged**
